### ISSUE TITLE
Ideas for E2E tests and peripherals

### ISSUE THREAD

**Issue by karuppiah7890** [karuppiah7890/tce-e2e-test#1]
- A report which shows what happened in the E2E test - the steps that happened, any actions, any verifications and which actions succeeded and which failed, which verifications passed and which failed, and any reason for failure, so that we can see quickly what's the problem! :) Instead of going through the full logs!! JSON report please!!

- Automated Golang program to pull GitHub Actions run data - which passed, which failed, logs of failed, ability to grep the logs and get some data etc to understand similarity etc, and get data like number of the test, name of the test (AWS, Azure etc) and the URL to the workflow, the jobs, the steps etc

**Comment by karuppiah7890** [comment]
When writing tests as native golang testing library tests, one big benefit we get is - we can choose which tests to run by choosing tests with regexp

**Comment by karuppiah7890** [comment]
When using credentials in E2E tests, the E2E test code should probably do a check on the credentials as the first thing - check if it's defined (as environment variable), not empty (empty string), and valid (not expired) and has appropriate permissions - enough permissions to run the E2E test, for example, create VMs, create networks, create disks, and similar so that it has enough permissions to create a cluster and destroy it too as part of the E2E test

**Comment by karuppiah7890** [comment]
When downloading tar ball for installation of `tanzu` from TCE or TF, we could show progress 😁 

https://duckduckgo.com/?t=ffab&q=show+progress+for+download+in+golang&ia=web

https://gist.github.com/albulescu/e61979cc852e4ee8f49c

We could use dots `.` to show progress, probably 10 or 20 dots to show 100 % progress. If it's 10 dots, each dot represents 10% being done, if it's 20 dots, each dot shows 5% progress. Why dots? So that it's CI/CD friendly and easy to print the dots in a line to show progress. Something like -

```
Progress:..........
```

**Comment by karuppiah7890** [comment]
Checking out existing E2E test framework in TCE and upstream - done by Shwetha and Vladimir and others
    - Take inspiration from it
    - Understand it
    - Understand pros and cons

https://github.com/vmware-tanzu/community-edition/tree/03201d8852c20c34a39086d4a113053bc990cdfd/test/e2e

https://github.com/kubernetes-sigs/e2e-framework

**Comment by karuppiah7890** [comment]
- Think about the requirements to satisfy
    - Providing libraries and functions to run clusters

**Comment by karuppiah7890** [comment]
For E2E tests, what we need to do is - show proper error messages. This is true for any software program actually. When something fails, try to show why exactly it fails and let it be as intelligent as possible and also give suggestions / advice on good action items on what to do next so that the user can take some actions to fix / get rid of the error

**Comment by karuppiah7890** [comment]
The vSphere E2E test can check if vSphere control plane endpoint(s) are already accessible before the test starts so that we can error out and say that there is an IP clash and that static virtual IP(s) that aren't in use only has to be provided as input. Note that `tanzu` CLI does this check too as part of it's validation and pre-requisites but we can do this way earlier as part of test input sanitization and fail early, unless we are testing this feature of `tanzu` CLI in which case we don't have to do this check in the test code and instead just check if `tanzu` CLI fails when using an IP that's already in use for a control plane endpoint IP

**Comment by karuppiah7890** [comment]
Test input for E2E tests - formats -
- CLI Flags and CLI flag values
- CLI arguments
- Environment variables
- Test config file

**Comment by karuppiah7890** [comment]
Have a `Makefile` for convenience and maybe use https://earthly.dev as the supported build system 😁

**Comment by karuppiah7890** [comment]
Apart from having a summary / report of what are the steps / actions that happened, we need to create logs for each of those steps and be able to look at the logs of a given step quickly. Note - the logs should mask all the test secret inputs. For example, if a log contains an error showing that the client secret passed is wrong / invalid and logs the client ID, we should mask it with dummy characters or a placeholder like "<client-id>"

Basically - mask all the test secret inputs / data with their corresponding names in a placeholder like fashion so that it's clear what was there but not clear about what's the exact value

**Comment by karuppiah7890** [comment]
Some sample logs for Azure test -

```bash
tce-e2e-test $ go test -v ./...
=== RUN   TestAzureManagementAndWorkloadCluster
2022/04/09 15:13:05 Checking required environment variables...
2022/04/09 15:13:05 Extracting Azure test secrets from environment variables...
2022/04/09 15:13:05 Logging into Azure...
2022/04/09 15:13:05 Getting marketplace terms for Azure VM image...
2022/04/09 15:13:11 Azure VM image agreement terms are already accepted
--- PASS: TestAzureManagementAndWorkloadCluster (5.68s)
PASS
ok  	github.com/karuppiah7890/tce-e2e-test	6.083s
tce-e2e-test $ go test -v ./...
=== RUN   TestAzureManagementAndWorkloadCluster
2022/04/09 15:13:21 Checking required environment variables...
2022/04/09 15:13:21 Extracting Azure test secrets from environment variables...
2022/04/09 15:13:21 Logging into Azure...
2022/04/09 15:13:21 Getting marketplace terms for Azure VM image...
2022/04/09 15:13:26 Azure VM image agreement terms are already accepted
--- PASS: TestAzureManagementAndWorkloadCluster (5.34s)
PASS
ok  	github.com/karuppiah7890/tce-e2e-test	5.665s
tce-e2e-test $ go test -v ./...
=== RUN   TestAzureManagementAndWorkloadCluster
2022/04/09 15:13:33 Checking required environment variables...
2022/04/09 15:13:33 Extracting Azure test secrets from environment variables...
2022/04/09 15:13:33 Logging into Azure...
2022/04/09 15:13:33 Getting marketplace terms for Azure VM image...
2022/04/09 15:13:39 Azure VM image agreement terms is not already accepted. Signing and accepting the Azure VM image agreement terms now...
2022/04/09 15:13:44 Error while signing and accepting the agreement terms for Azure VM image: POST https://management.azure.com/subscriptions/87fbb054-796e-47cc-a777-c7a052901de1/providers/Microsoft.MarketplaceOrdering/agreements/vmware-inc/offers/tkg-capi/plans/k8s-1dot22dot8-ubuntu-2004/sign
--------------------------------------------------------------------------------
RESPONSE 400: 400 Bad Request
ERROR CODE: BadRequest
--------------------------------------------------------------------------------
{
  "error": {
    "code": "BadRequest",
    "message": "You must accept the legal terms through the Azure portal (http://go.microsoft.com/fwlink/?LinkId=534873). Correlation Id: '479de9c1-f4e8-4879-a573-82d6a8399699'."
  }
}
--------------------------------------------------------------------------------
FAIL	github.com/karuppiah7890/tce-e2e-test	10.769s
FAIL
tce-e2e-test $ 
```

**Comment by karuppiah7890** [comment]
https://docs.microsoft.com/en-us/rest/api/marketplaceordering/marketplace-agreements/create

https://docs.microsoft.com/en-us/rest/api/marketplaceordering/marketplace-agreements/sign

**Comment by karuppiah7890** [comment]
```go
signResponse, err := client.Sign(ctx, azureVmImagePublisher,
	azureVmImageOffer,
	azureVmImageBillingPlanSku,
	nil)
if err != nil {
	log.Fatalf("Error while signing and accepting the agreement terms for Azure VM image: %+v", err)
}

signedAgreementTerms := signResponse.AgreementTerms

if signedAgreementTerms.Properties == nil {
	log.Fatalf("Error while signing and accepting the agreement terms for Azure VM image: Azure VM image agreement terms Properties field is not available")
}

if signedAgreementTerms.Properties.Accepted == nil {
	log.Fatalf("Error while signing and accepting the agreement terms for Azure VM image: Azure VM image agreement terms Properties Accepted field is not available")
}

if isTermsSignedAndAccepted := *signedAgreementTerms.Properties.Accepted; !isTermsSignedAndAccepted {
	log.Fatalf("Error while signing and accepting the agreement terms for Azure VM image: Azure VM image agreement terms was not signed and accepted")
}
```

**Comment by karuppiah7890** [comment]
```bash
tce-e2e-test $ go test -v ./...
=== RUN   TestAzureManagementAndWorkloadCluster
2022/04/09 16:26:37 Checking required environment variables...
2022/04/09 16:26:37 Extracting Azure test secrets from environment variables...
2022/04/09 16:26:37 Logging into Azure...
2022/04/09 16:26:37 Getting marketplace terms for Azure VM image...
2022/04/09 16:26:42 Azure VM image agreement terms is not already accepted. Signing and accepting the Azure VM image agreement terms now...
2022/04/09 16:26:46 Error while signing and accepting the agreement terms for Azure VM image: POST https://management.azure.com/subscriptions/87fbb054-796e-47cc-a777-c7a052901de1/providers/Microsoft.MarketplaceOrdering/agreements/vmware-inc/offers/tkg-capi/plans/k8s-1dot22dot8-ubuntu-2004/sign
--------------------------------------------------------------------------------
RESPONSE 400: 400 Bad Request
ERROR CODE: BadRequest
--------------------------------------------------------------------------------
{
  "error": {
    "code": "BadRequest",
    "message": "You must accept the legal terms through the Azure portal (http://go.microsoft.com/fwlink/?LinkId=534873). Correlation Id: 'ccd645a2-389a-4157-8624-43d597d59743'."
  }
}
--------------------------------------------------------------------------------
FAIL	github.com/karuppiah7890/tce-e2e-test	9.753s
FAIL
tce-e2e-test $ 
```

https://duckduckgo.com/?t=ffab&q=You+must+accept+the+legal+terms+through+the+Azure+portal&ia=web

https://www.nblabs.net/2017/12/20/accepting-azure-marketplace-terms-powershell/

https://carlos.mendible.com/2018/05/21/accepting-azure-marketplace-terms-with-azure-cli/

https://duckduckgo.com/?t=ffab&q=azure+cli&ia=web

https://github.com/Azure/azure-cli

https://github.com/Azure/azure-cli/search?q=accept-terms

https://github.com/Azure/azure-cli/blob/9bae6080ec1fc258ffb52f7218915971f6e8051f/src/azure-cli/azure/cli/command_modules/vm/commands.py

https://github.com/Azure/azure-cli/blob/9bae6080ec1fc258ffb52f7218915971f6e8051f/src/azure-cli/azure/cli/command_modules/vm/commands.py#L390

https://github.com/Azure/azure-cli/blob/9bae6080ec1fc258ffb52f7218915971f6e8051f/src/azure-cli/azure/cli/command_modules/vm/commands.py#L394

https://github.com/Azure/azure-cli/blob/9bae6080ec1fc258ffb52f7218915971f6e8051f/src/azure-cli/azure/cli/command_modules/vm/commands.py#L111-L114

`azure.mgmt.marketplaceordering.operations#MarketplaceAgreementsOperations`

**Comment by karuppiah7890** [comment]
https://docs.microsoft.com/en-us/answers/questions/52637/cannot-sign-azure-marketplace-vm-image-licence-thr.html

**Comment by karuppiah7890** [comment]
```bash

tce-e2e-test $ go clean -testcache
tce-e2e-test $ go test -v ./...
=== RUN   TestAzureManagementAndWorkloadCluster
2022/04/09 17:06:36 Checking required environment variables...
2022/04/09 17:06:36 Extracting Azure test secrets from environment variables...
2022/04/09 17:06:36 Logging into Azure...
2022/04/09 17:06:36 Getting marketplace terms for Azure VM image...
2022/04/09 17:06:42 Azure VM image agreement terms is not already accepted. Accepting the Azure VM image agreement terms now...
2022/04/09 17:06:47 Accepted the Azure VM image agreement terms!
--- PASS: TestAzureManagementAndWorkloadCluster (10.93s)
PASS
ok  	github.com/karuppiah7890/tce-e2e-test	11.305s
tce-e2e-test $ az vm image terms show --publisher "vmware-inc" --offer "tkg-capi"         --plan "k8s-1dot22dot8-ubuntu-2004" --subscription "${AZURE_SUBSCRIPTION_ID}" --debug
{
  "accepted": true,
  "id": "/subscriptions/<subscription-id>/providers/Microsoft.MarketplaceOrdering/offerTypes/VirtualMachine/publishers/vmware-inc/offers/tkg-capi/plans/k8s-1dot22dot8-ubuntu-2004/agreements/current",
  "licenseTextLink": "https://mpcprodsa.blob.core.windows.net/legalterms/3E5ED_legalterms_VMWARE%253a2DINC%253a24TKG%253a2DCAPI%253a24K8S%253a2D1DOT22DOT8%253a2DUBUNTU%253a2D2004%253a24KQJXKK2NIUVNWR2B7X4J2ZWNZGFLGIQPBB46SSI4WO32VUTVGS6WFNL5JR3J6WPKJZK5YCCZV7WEHQGADEWGPC7PEJHUIA2S4ONKRFY.txt",
  "marketplaceTermsLink": "https://mpcprodsa.blob.core.windows.net/marketplaceterms/3EDEF_marketplaceterms_VIRTUALMACHINE%253a24AAK2OAIZEAWW5H4MSP5KSTVB6NDKKRTUBAU23BRFTWN4YC2MQLJUB5ZEYUOUJBVF3YK34CIVPZL2HWYASPGDUY5O2FWEGRBYOXWZE5Y.txt",
  "name": "k8s-1dot22dot8-ubuntu-2004",
  "plan": "k8s-1dot22dot8-ubuntu-2004",
  "privacyPolicyLink": "https://www.vmware.com/help/privacy/products-and-services-notice.html",
  "product": "tkg-capi",
  "publisher": "vmware-inc",
  "retrieveDatetime": "2022-04-09T11:37:01.5262201Z",
  "signature": "QN2LE7QN5TSE57PKJIAYN326FVPAWUS4SRBYUANY4W7ETANMMDORKWXSQ543VP7YKQTI2EAI2ESHH6IAPE2R34WDWTUZTCAUVDG5I5Q",
  "systemData": {
    "createdAt": "2022-04-09T11:37:01.598348+00:00",
    "createdBy": "<subscription-id>",
    "createdByType": "ManagedIdentity",
    "lastModifiedAt": "2022-04-09T11:37:01.598348+00:00",
    "lastModifiedBy": "<subscription-id>",
    "lastModifiedByType": "ManagedIdentity"
  },
  "type": "Microsoft.MarketplaceOrdering/offertypes"
}
```

**Comment by karuppiah7890** [comment]
https://mpcprodsa.blob.core.windows.net/legalterms/3E5ED_legalterms_VMWARE%253a2DINC%253a24TKG%253a2DCAPI%253a24K8S%253a2D1DOT22DOT8%253a2DUBUNTU%253a2D2004%253a24KQJXKK2NIUVNWR2B7X4J2ZWNZGFLGIQPBB46SSI4WO32VUTVGS6WFNL5JR3J6WPKJZK5YCCZV7WEHQGADEWGPC7PEJHUIA2S4ONKRFY.txt

https://mpcprodsa.blob.core.windows.net/marketplaceterms/3EDEF_marketplaceterms_VIRTUALMACHINE%253a24AAK2OAIZEAWW5H4MSP5KSTVB6NDKKRTUBAU23BRFTWN4YC2MQLJUB5ZEYUOUJBVF3YK34CIVPZL2HWYASPGDUY5O2FWEGRBYOXWZE5Y.txt

Examples of `.env` files - https://terraspace.cloud/docs/config/dotenv/

**Comment by karuppiah7890** [comment]
Before running tests against management and workload cluster or any plugin, we can check if the plugin exists and is accessible like this -

```bash
$ tanzu management-cluster version
v0.11.2

$ tanzu cluster version
v0.11.2

$ tanzu unmanaged-cluster version
v0.11.0
```

**Comment by karuppiah7890** [comment]
We can parse the version of the plugins and show warning in case the E2E tests have not been tried and tested with newer versions. We can hard code the version we have tried and tested it with and anything (any version) newer - it's possible to have breaking changes etc, which may result in the test not working correctly to test the newer version

**Comment by karuppiah7890** [comment]
The E2E tests can fail early if any of the required plugins are not present for doing the current test. For example, if any of management and workload cluster plugins are not there, they cannot run management and workload cluster plugin test

**Comment by karuppiah7890** [comment]
The E2E tests can fail early by checking if Docker daemon is running and accessible. This way it's faster for it to fail. But yeah, Tanzu CLI already does this and fails, but it's way later in the process of the E2E test but not too late I guess as it's the first step in Tanzu CLI, to check pre-requisites

**Comment by karuppiah7890** [comment]
Examples for listing workload clusters -

```bash
$ tanzu cluster list -o json
[
  {
    "name": "workload-cluster-26-08-2021-1",
    "namespace": "default",
    "status": "deleting",
    "plan": "dev",
    "controlplane": "0/1",
    "workers": "",
    "kubernetes": "v1.21.2+vmware.1",
    "roles": [],
    "labels": {
      "tanzuKubernetesRelease": "v1.21.2---vmware.1-tkg.1-zshippable",
      "tkg.tanzu.vmware.com/cluster-name": "workload-cluster-26-08-2021-1"
    }
  }
]

$ tanzu cluster list
  NAME                           NAMESPACE  STATUS    CONTROLPLANE  WORKERS  KUBERNETES        ROLES   PLAN  
  workload-cluster-26-08-2021-1  default    deleting  0/1                    v1.21.2+vmware.1  <none>  dev   

$ tanzu cluster list
  NAME  NAMESPACE  STATUS  CONTROLPLANE  WORKERS  KUBERNETES  ROLES  PLAN
```

**Comment by karuppiah7890** [comment]
https://pkg.go.dev/bytes - for `bytes.Buffer`

https://pkg.go.dev/encoding/json - for JSON decoding

**Comment by karuppiah7890** [comment]
Store all command logs (standard output, standard err, combined - just for easy use) in log files and ensure that the secrets are masked in these log files. Note - let's not post process these files to mask the secrets, instead, when we write the log to the files itself the secrets should be masked

```

                      Secrets
                        ||
                        ||
                        \/
output stream => secrets masker => write masked output to file
```

**Comment by karuppiah7890** [comment]
The test should know it's timeout time and accordingly ensure that the test is over within that time. Or else, the test can start cleaning up things. For example, if the timeout time is 30 minutes, the test can have a max time of 20 minutes and then trigger off a cancellation / stop / cleanup of all the things it started off so that they will be done in those 10 minutes. Or, we can do the cleanup separately but we need to know what is half way done / half baked to do the cleanup, or know what actions started and happened to revert them

**Comment by karuppiah7890** [comment]
A logger library which we can use to log at different levels including custom levels (summary etc) and can log to different destinations - standard output, standard error, log files, summary files etc and also log in structured format (json etc) and unstructured format (simple log line)

We can consider third part log libraries. I have heard Uber Zap to be popular - https://github.com/uber-go/zap/

**Comment by karuppiah7890** [comment]
At any point in time, we should be able to understand the progress of the E2E test. Use progress bars / dots etc, use logs etc to show that there is some progress and how much progress is there - some sort of percentage apart from indefinite progress notification

We should be able to find out what the E2E test is doing at any point in time and be able to tell if it's stuck or if it's just taking time to finish it's job

**Comment by karuppiah7890** [comment]
Idea for Credentials - Login to the cloud account exactly once and reuse that among the test functions. Either pass the credentials all over, or use a package which stores an inaccessible variable and provides it to globally as a singleton. Logic? If not already available in the variable, instantiate and store it in the single variable and return. If already available, just return

Idea for log - same as credentials, provide it globally as a singleton through a package. Logic? If not already available in the variable, instantiate and store it in the single variable and return. If already available, just return

**Comment by karuppiah7890** [comment]
We gotta check required quota and quota available before running E2E tests to avoid errors like the below -

```
controller/azuremachine "msg"="Reconciler error" "error"="failed to reconcile AzureMachine: failed to create virtual machine: failed to create VM test-wkld-1649820052-control-plane-gsfj8 in resource group test-wkld-1649820052: compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code=\"OperationNotAllowed\" Message=\"Operation could not be completed as it results in exceeding approved Total Regional Cores quota. Additional details - Deployment Model: Resource Manager, Location: australiaeast, Current Limit: 10, Current Usage: 8, Additional Required: 4, (Minimum) New Limit Required: 12. Submit a request for Quota increase at https://aka.ms/ProdportalCRP/#blade/Microsoft_Azure_Capacity/UsageAndQuota.ReactView/Parameters/%7B%22subscriptionId%22:%22<subscription-id>%22,%22command%22:%22openQuotaApprovalBlade%22,%22quotas%22:[%7B%22location%22:%22australiaeast%22,%22providerId%22:%22Microsoft.Compute%22,%22resourceName%22:%22cores%22,%22quotaRequest%22:%7B%22properties%22:%7B%22limit%22:12,%22unit%22:%22Count%22,%22name%22:%7B%22value%22:%22cores%22%7D%7D%7D%7D]%7D by specifying parameters listed in the ‘Details’ section for deployment to succeed. Please read more about quota limits at https://docs.microsoft.com/en-us/azure/azure-supportability/regional-quota-requests\"" "name"="test-wkld-1649820052-control-plane-gsfj8" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AzureMachine" 
```
